Trivial change arf

ezyang opened the pull request:
Trivial change arf

Summary: none

Differential Revision: https://phabricator.intern.facebook.com/D9514646

ezyang commented:
Thanks! My high level:

* I know I wasn't very convinced by the operator/instruction/expr collapsing when we talked about it earlier, but looking at it now it looks good to me. :+1: 

* I'd prefer the pretty-printer to not be implemented using overloaded `<<`. Keeping it separate gives us the flexibility to add an indentation, precedence or a generalized settings parameter. I don't think this is premature generalization: in `pointwise-fusion`, I've been planning to add back indentation so that nested fragments of graphs render properly. This is a planned feature in ToffeeIR.

* I'm not a big fan of the name `Value`. In the greek-letter PL world, value exclusively refers to "term which cannot reduce further (sample citation: https://softwarefoundations.cis.upenn.edu/current/Smallstep.html#lab492) and variables definitely do not have this property. Nor does `Value` as a term coincide with any current ToffeeIR terminology. I am not an LLVM expert, but I guess the term `Value` comes from LLVM? But LLVM takes a far more expansive view about values: a value is anything that may be used as an operand to an instruction. That is not the function of values/locals in the IR today: they really are just variables (but we can't call them variables because PyTorch has already taken that name.) You can't "bind" a value in the inputs of a graph, that doesn't make sense, even in LLVM. So the naming here is something of a faux amis.

* Can we bring back the visitor for Op? I really don't want to encourage a coding style where every time you do an exhaustive case, you have to (1) remember to use `__builtin_unreachable()` (in the patch you throw, which is bad, because you've turned a compile time exception into a runtime exception) and (2) have to manually cast (`evalPythonOp((PythonOp*)i.get(), vars)`): anywhere we do a manual cast is code that needs to be inspected carefully, and I'd prefer to minimize the TCB here by putting it in a visitor.

ezyang commented:
CC @apaszke

zdevito commented:
I added a commit that does refcounting of Node/Value. NodeRef/ValueRef are strong shared pointers that will keep the Node or Value alive that they refer to.  Values do not individually have a refcount, instead they forward their retain/release to the Node that defines them. The Node owns the actual Value objects that are its outputs via unique_ptrs. This ensures that a Node+its output Values stay alive as long as there is either a reference directly to a Node, or a reference to Value that is defined by that node. When there are no more references, the node is delete, which causes the output Values to be deleted as well. I still want to add helper functions that initialize outputs because it is necessary for this initialization process to understand this ownership scheme.

ezyang commented on torch/csrc/autograd/ir.h:
```diff
@@ -127,12 +138,29 @@ struct Node {
+  friend class torch::IntrusivePtr<Value>;
+  friend class torch::IntrusivePtr<Node>;
+  friend class Value;
+  std::atomic<uint32_t> refcount_;
```
It would be good to have a comment explaining why shared_ptr is not suitable here!

ezyang replied:
Yeah, I think we can do this with shared_ptr?

What are the constraints? As far as I can tell:

1. A Value is owned by the Node that defines them,
2. A reference to a Value keeps the Node alive
3. The owning reference from the Node does NOT count

So you could reuse shared_ptr by setting up the memory management in this way:

* Value holds shared_ptr reference to Node
* Node holds non-smart reference to Value. However, this is an "owning" reference: Node's constructor is responsible for allocating Values and destructor is responsible for deallocating them. I guess we could make up a special "smart" pointer for indicating this, which is just a wrapper on regular pointer.

ezyang replied:
It seems to me that the intrusive pointer is an optimization over an alternate strategy: define `NodeRef` as the following:

```
struct NodeRef {
  std::shared_ptr<Node>,
  int offset;
}
```

As far as I can tell, this fulfills the constraints above. It's also basically what is implemented in autograd today.

apaszke replied:
It's not equivalent. We want two-way ownership, and your code would create a ref-cycle

ezyang replied:
How would it create a cycle? To be clear, in this regime, Node has a unique pointer to a Value, a Value has a non-smart pointer to a Node.

apaszke replied:
Oh sorry, I misunderstood what you wrote. In fact what you propose is exactly what we use in autograd right now (except that we use `std::pair` instead of a `NodeRef`)

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -706,15 +709,15 @@ static std::shared_ptr<Node> make_trace(PyObject* pyobj, PyObject *arg_objects,
     }
     Py_INCREF(pyobj);
-    auto op = std::make_shared<PythonOp>(
+    auto op = NodeRef(new PythonOp(
```
Can we add a `make_ref<PythonOp>()` helper function for uniformity with `make_shared`? This way, we can keep use of `new` as a code smell (it means you're doing something wrong.)

ezyang commented on torch/csrc/autograd/ir.h:
```diff
@@ -89,19 +91,28 @@ struct Node;
-  Value(int unique)
-    : unique(unique)
-    {}
+  Node * definition;
```
I would really like a comment explaining the ownership structure of the AST now. Maybe something like what you wrote in a comment on the PR:

> NodeRef/ValueRef are strong shared pointers that will keep the Node or Value alive that they refer to. Values do not individually have a refcount, instead they forward their retain/release to the Node that defines them. The Node owns the actual Value objects that are its outputs via unique_ptrs. This ensures that a Node+its output Values stay alive as long as there is either a reference directly to a Node, or a reference to Value that is defined by that node. When there are no more references, the node is delete, which causes the output Values to be deleted as well.

ezyang replied:
For example, if you make use of the `definition` pointer, you MUST keep the originating `ValueRef` alive, despite the fact that the ValueRef does not "own" the definition pointer.

zdevito replied:
This commit was about figuring out how to implement the correct memory management. The yet-to-be-written accessor function that returns the definition can return a NodeRef.

zdevito commented:
Given our discussion earlier, the IR is going to change a bit now that it doesn't have to represent running code. I'll send another patch tomorrow with what it will look like, but a lot of things like ownership can now be simplified.

apaszke commented:
How about a VC tomorrow? The new constraints are completely different, so it might be worth having that discussion once again

zdevito commented:
I pushed a new revision of this and merged it in, so that we have branch where we can start working next week. Documentation is inline.

Last time we met, I advocated copying LLVM/Swift IR for how mutation and use-def works. Looking at their actual code (https://github.com/apple/swift/blob/master/include/swift/SIL/SILValue.h, see Operand in particular), it feels unnecessarily complicated for what we need. Lots of the tracking code creates weird container types that then permeate everything else.

Instead of going down this route, I want to take Ed's approach of keeping the Graph/Nodes functional. While each provides methods to help build it incrementally (like adding inputs and new nodes), once constructed the nodes/graph are immutable. This keeps use-def tracking and all the container types simple. It is possible this will make transformation passes too complicated, but I am willing to try the functional way first. Adding mutation to this is always possible later if we really need it, but as Ed noted the reverse is not possible.

ezyang commented:
:( That's too bad, I was looking forward to seeing a miniature LLVM/Swift style IR for our project. If you think you need more time to work out how to do this (I probably spent a week hemming and hawing over the functional IR) I'm still OK with shifting at a later point.

ezyang committed:
Trivial change arf

Summary: none

Differential Revision: https://phabricator.intern.facebook.com/D9514646
